lib/charms/synapse/v0/matrix_auth.py

```diff
@@ -0,0 +1,512 @@
+            registration_secret_id = (
+                (relation_data["registration_secret_id"])
+                if "registration_secret_id" in relation_data
+                else None
+            )
+            registration = MatrixAuthRequirerData.get_registration(model, registration_secret_id)
+            return MatrixAuthRequirerData(
+                registration=registration,
+            )
+
+        except json.JSONDecodeError as ex:
+            logger.warning("Invalid relation data %s", ex)
+            raise ValueError from ex
+
+
+#### Events ####
+class MatrixAuthRequestProcessed(ops.RelationEvent):
+    """MatrixAuth event emitted when a new request is processed."""
+
+    def get_matrix_auth_provider_relation_data(self) -> MatrixAuthProviderData:
+        """Get a MatrixAuthProviderData for the relation data.
+
+        Returns:
+            the MatrixAuthProviderData for the relation data.
+        """
+        return MatrixAuthProviderData.from_relation(self.model, self.relation)
+
+
+class MatrixAuthRequestReceived(ops.RelationEvent):
+    """MatrixAuth event emitted when a new request is made."""
+
+
+class MatrixAuthRequiresEvents(ops.CharmEvents):
+    """MatrixAuth requirer events.
+
+    This class defines the events that a MatrixAuth requirer can emit.
+
+    Attributes:
+        matrix_auth_request_processed: the MatrixAuthRequestProcessed.
+    """
+
+    matrix_auth_request_processed = ops.EventSource(MatrixAuthRequestProcessed)
+
+
+class MatrixAuthProvidesEvents(ops.CharmEvents):
+    """MatrixAuth provider events.
+
+    This class defines the events that a MatrixAuth provider can emit.
+
+    Attributes:
+        matrix_auth_request_received: the MatrixAuthRequestReceived.
+    """
+
+    matrix_auth_request_received = ops.EventSource(MatrixAuthRequestReceived)
+
+
+#### Requires and Provides ####
+class MatrixAuthRequires(ops.Object):
+    """Requirer side of the MatrixAuth requires relation.
+
+    Attributes:
+        on: events the provider can emit.
+    """
+
+    on = MatrixAuthRequiresEvents()
+
+    def __init__(self, charm: ops.CharmBase, relation_name: str = DEFAULT_RELATION_NAME) -> None:
+        """Construct.
+
+        Args:
+            charm: the provider charm.
+            relation_name: the relation name.
+        """
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.relation_name = relation_name
+        self.framework.observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+
+    def get_remote_relation_data(self) -> Optional[MatrixAuthProviderData]:
+        """Retrieve the remote relation data.
+
+        Returns:
+            MatrixAuthProviderData: the relation data.
+        """
+        relation = self.model.get_relation(self.relation_name)
+        return self._get_remote_relation_data(relation) if relation else None
+
+    def _get_remote_relation_data(self, relation: ops.Relation) -> MatrixAuthProviderData:
```

Comment:
Is it necessary to have a method for a one liner ?

Reply:
Closing this one since this library is hosted in the synapse repo.